File: app/src/main/java/com/example/recipeclipper/data/model/ShortStepCheck.kt
package com.example.recipeclipper.data.model

/**
 * Chef mode's gate (#100): whether a short version of a step, written by the on-device model,
 * may be shown in place of the step as written. Pure, and the same on iOS (the differential
 * corpus's `Short` rows pin it).
 *
 * The model writes words; code owns every number. A short version passes only if it is shorter
 * than the step, every number in it (fractions, decimal commas and each end of a range, as
 * written) appears in the step, and it states exactly the step's times and temperatures: none
 * changed, none added, none dropped ("350°F (180°C)" may keep either half). Anything else, and
 * any recipe language the app has no words for, shows the step as written.
 */
object ShortStepCheck {

    /** [short], tidied, when it may stand for [original]; null to show [original] as written. */
    fun accept(original: String, short: String?, words: LanguageWords?): String? {
        if (words == null || short == null) return null
        val candidate = tidy(short)
        if (candidate.isEmpty() || candidate.length >= tidy(original).length) return null
        if (!numbers(original).containsAll(numbers(candidate))) return null
        if (StepTimers.durations(original, words).toSet() != StepTimers.durations(candidate, words).toSet()) {
            return null
        }
        val stated = TemperatureConverter.temperatures(original, words)
        val kept = TemperatureConverter.temperatures(candidate, words).flatten().toSet()
        if (!stated.flatten().containsAll(kept) || stated.any { keys -> keys.none { it in kept } }) return null
        return candidate
    }

    /** A step this short ("Serve warm.") is left as written, never sent to the model. */
    fun worthShortening(step: String): Boolean = tidy(step).length >= MIN_LENGTH

    const val MIN_LENGTH = 40

    /** One line: trimmed, whitespace collapsed, a leading bullet and wrapping quotes removed. */
    fun tidy(text: String): String {
        var s = text.trim().replace(SPACES, " ")
        s = s.replace(BULLET, "")
        if (s.length >= 2 && s.first() in QUOTES && s.last() in QUOTES) s = s.substring(1, s.length - 1).trim()
        return s
    }

    /** Every number in [text], as written: "1,5", "1 1/2", "1½", "½"; a range's ends separately. */
    fun numbers(text: String): Set<String> {
        val plain = buildString {
            for (c in text) {
                append(
                    when (c) {
                        in '０'..'９' -> '0' + (c - '０')
                        '／', '⁄' -> '/'
                        else -> c
                    }
                )
            }
        }
        return NUMBER.findAll(plain).map { it.value.replace(SPACES, " ").replace(SPACED_FRACTION, "$1$2") }.toSet()
    }

    private const val FRACTIONS = "½⅓⅔¼¾⅕⅖⅗⅘⅙⅚⅐⅛⅜⅝⅞⅑⅒"
    private val NUMBER = Regex("""\d+(?:[.,]\d+)*(?:\s+\d+/\d+|/\d+|\s*[$FRACTIONS])?|[$FRACTIONS]""")
    private val SPACED_FRACTION = Regex("""(\d) ([$FRACTIONS])""")
    private val SPACES = Regex("""\s+""")
    private val BULLET = Regex("""^[-•*]\s+""")
    private const val QUOTES = "\"'“”„«»「」"
}

File: ios/RecipeClipper/Data/Model/ShortStepCheck.swift
import Foundation

/// Chef mode's gate (#100): whether a short version of a step, written by the on-device model,
/// may be shown in place of the step as written. Pure, and the same as Android's (the
/// differential corpus's `Short` rows pin it).
///
/// The model writes words; code owns every number. A short version passes only if it is shorter
/// than the step, every number in it (fractions, decimal commas and each end of a range, as
/// written) appears in the step, and it states exactly the step's times and temperatures: none
/// changed, none added, none dropped ("350°F (180°C)" may keep either half). Anything else, and
/// any recipe language the app has no words for, shows the step as written.
enum ShortStepCheck {

    /// `short`, tidied, when it may stand for `original`; nil to show `original` as written.
    static func accept(_ original: String, _ short: String?, words: LanguageWords?) -> String? {
        guard let words, let short else { return nil }
        let candidate = tidy(short)
        // UTF-16 lengths, as Kotlin's String.length.
        if candidate.isEmpty || candidate.utf16.count >= tidy(original).utf16.count { return nil }
        if !numbers(candidate).isSubset(of: numbers(original)) { return nil }
        if Set(StepTimers.durations(original, words: words)) != Set(StepTimers.durations(candidate, words: words)) {
            return nil
        }
        let stated = TemperatureConverter.temperatures(original, words: words)
        let kept = Set(TemperatureConverter.temperatures(candidate, words: words).flatMap { $0 })
        if !kept.isSubset(of: Set(stated.flatMap { $0 })) || stated.contains(where: { $0.allSatisfy { !kept.contains($0) } }) {
            return nil
        }
        return candidate
    }

    /// A step this short ("Serve warm.") is left as written, never sent to the model.
    static func worthShortening(_ step: String) -> Bool { tidy(step).utf16.count >= minLength }

    static let minLength = 40

    /// One line: trimmed, whitespace collapsed, a leading bullet and wrapping quotes removed.
    static func tidy(_ text: String) -> String {
        var s = spaces.replace(text.kTrimmed, with: " ")
        s = bullet.replace(s, with: "")
        if s.utf16.count >= 2, let first = s.first, let last = s.last, quotes.contains(first), quotes.contains(last) {
            s = String(s.dropFirst().dropLast()).kTrimmed
        }
        return s
    }

    /// Every number in `text`, as written: "1,5", "1 1/2", "1½", "½"; a range's ends separately.
    static func numbers(_ text: String) -> Set<String> {
        var plain = ""
        for scalar in text.unicodeScalars {
            switch scalar.value {
            case 0xFF10...0xFF19: plain.unicodeScalars.append(Unicode.Scalar(scalar.value - 0xFF10 + 0x30)!)
            case 0xFF0F, 0x2044: plain.append("/")
            default: plain.unicodeScalars.append(scalar)
            }
        }
        return Set(number.findAll(plain).map { spacedFraction.replace(spaces.replace($0.value, with: " ")) { "\($0[1])\($0[2])" } })
    }

    private static let fractions = "½⅓⅔¼¾⅕⅖⅗⅘⅙⅚⅐⅛⅜⅝⅞⅑⅒"
    private static let number = JRegex(#"\d+(?:[.,]\d+)*(?:\s+\d+/\d+|/\d+|\s*["# + fractions + #"])?|["# + fractions + "]")
    private static let spacedFraction = JRegex(#"(\d) (["# + fractions + "])")
    private static let spaces = JRegex(#"\s+"#)
    private static let bullet = JRegex(#"^[-•*]\s+"#)
    private static let quotes: Set<Character> = ["\"", "'", "“", "”", "„", "«", "»", "「", "」"]
}
